Initial Implementation for Wasm Backend

Derppening opened the pull request:
This pull request implements rudimentary support for compiling MLx programs into WebAssembly's text format (WAT).

Notes:

- `WasmGenerator.scala` and `BinaryenJSBackend.scala` are not needed in the actual implementation, and can be deleted if not useful.
- The contents of the `wasm` virtual module is not really useful AFAIK.

Derppening committed:
Minor cleanup

Derppening committed:
WatBackend: Fix incorrect instr for nop

Derppening committed:
Fix several missing newlines/brackets in outputted WAT

Derppening committed:
BinaryenJSBackend: Implement mod.{ref,i31ref}

Derppening committed:
Add module.call

Derppening committed:
WatBackend: Mutate module in-place

Derppening committed:
WatBackend: Flatten ParamLists when determining function params

Derppening committed:
WatBackend: Add temp hardcoded handling of direct function calls

Derppening committed:
BinaryenJSBackend: Add close for dropping dmoules after use

Derppening committed:
Add Module.{Ref.func,callRef}

Derppening committed:
WatBackend: Replace ??? with TODO

Derppening committed:
WatBackend: Emit calls as call_ref

Derppening committed:
Refactor WasmGenerator

We need types to be a generic in a later commit. WIP - Needs to remove
redundant new expr

Derppening committed:
WatBackend: Implement dynamic function resolution

Derppening committed:
Implement getExpressionType for instructions

Derppening committed:
Move WasmType to WatBackend

Not used by Binaryen backend.

Derppening committed:
WatBackend: Fix incorrect instruction types

Derppening committed:
WIP - Add getExprType

Derppening committed:
Fill in correct type for ref.func

Derppening committed:
Move WasmType to Module.scala

WasmType is backend-independent.

Derppening committed:
Fix I31RefType text representation

Derppening committed:
Add WasmGenerator.getFunction

Allows getting a handle to an existing function.

Derppening committed:
Add WasmGenerator.getFunctionInfo

For obtaining the param/result types.

Derppening committed:
Add TODOs

Derppening committed:
BinaryenJSBackend: Add explicit semicolons

Derppening committed:
BinaryenJSBackend: Pre-define all variables

Derppening committed:
WIP: Initial commit for TypeBuilder and SignatureType

Derppening committed:
Add bitor to RuntimeJS

Derppening committed:
WatBackend: Use inferred expr type for main fn return type

Derppening committed:
Fix lowering of unreachable types to WAT

Derppening committed:
WIP: Initial commit for testing use of module functions

Derppening committed:
Lowering: Add wasm as virtual module

Derppening committed:
Add {bitand,bitnot,shl} to Runtime.mls

Derppening committed:
WIP: Initial commit for Int31 type

Derppening committed:
WIP: Implement if instruction

Untested for now.

Derppening committed:
WIP: Use i32 by default and cast type on use

Derppening committed:
Implement support for boolean literals

Derppening committed:
Implement ref.cast

Derppening committed:
Add type casts to add implementation

Derppening committed:
WIP: Add compilation target

Derppening committed:
Add binaryen to npm dependencies

We use nightly since it contains GC APIs.

Derppening committed:
Initial commit for WasmGenerator, WatBackend, BinaryenJSBackend

Derppening committed:
Implement addition operator for i32

Derppening committed:
Add WasmDiffMaker, Wasm.mls utilities

Derppening committed:
Implement support for packed types

Used in structure types.

Derppening committed:
Add TypeBuilder.setStructType

Derppening committed:
Simplify usage of `TypeBuilder.setStructType`

Derppening committed:
WIP: Add Fields.apply docs

Derppening committed:
Apply scalafmt

Derppening committed:
Remove all unnecessary `override` keywords

Derppening committed:
Fix missing newline

Derppening committed:
Module: Introduce RefType and convert all reftypes to HeapType

Derppening committed:
Add struct.new instruction

Derppening committed:
Remove StringRef

This is not part of the typed ref+GC+tail call spec.

Derppening committed:
WatBackend: Unify WAT output functions

Now all under the ToWat trait.

Derppening committed:
Module: Add ModType

Derppening committed:
WatBackend: Fixup struct.new

Add correct type to instruction.

Derppening committed:
Convert all ??? to WarningReport

Derppening committed:
WIP for compilation of classes

Derppening committed:
WIP: Add struct layout of class to module

Derppening committed:
Convert to use raw function names

Derppening committed:
Add struct.new_default

Derppening committed:
WIP: Add TypeRef to Module, return TypeRefs when adding types, emit
correct instruction

Derppening committed:
Merge remote-tracking branch 'upstream/hkmc2' into wasm-backend

Derppening committed:
Fix missing changes to diff tests

Derppening committed:
Fix incorrect handling of ClsLikeDefn

Derppening committed:
Simplify generation of `(ref null func)`

Derppening committed:
Merge remote-tracking branch 'upstream/hkmc2' into wasm-backend

Derppening committed:
WatBackend: Fix call_ref instruction format

Derppening committed:
WIP - Resolve `<class>.class` as struct creation

Derppening committed:
Merge remote-tracking branch 'upstream/hkmc2' into wasm-backend

Derppening committed:
Add TODO for class name mangling

Derppening committed:
WatBackend: Implement result() of Select on class

Derppening committed:
WatBackend: Make TypeRef a heap type

Allows the use of type IDs in `(ref ...)` instead of forcibly expanding
everything.

Derppening committed:
WatBackend: Fix class name in not-implemented warning messages

Derppening committed:
Module: Add missing ')'

Derppening committed:
WatBackend: Implement constructor definition generation

Derppening committed:
Merge remote-tracking branch 'upstream/hkmc2' into wasm-backend

Derppening committed:
WatBackend: Fix WAT formatting for block result type

Derppening committed:
WasmGenerator: Initial commit for `local.{get,set}`

Derppening committed:
WatBackend: WIP move object creation into ctor

Derppening committed:
Merge remote-tracking branch 'upstream/hkmc2' into wasm-backend

Derppening committed:
Apply scalafmt changes

Derppening committed:
WatBackend: Fix passing i32 params to functions

Derppening committed:
WatBackend: Implement local.{get,set} instructions

Derppening committed:
Merge remote-tracking branch 'upstream/hkmc2' into wasm-backend

Derppening committed:
WatBackend: Add struct field identifier to WAT

Derppening committed:
Add missing line break

Dumb.

Derppening committed:
WIP: Implement `this` resolution

Derppening committed:
Remove redundant match branch

Derppening committed:
WIP: Continue implementing `this` resolution and locals

Derppening committed:
Add known good Wasm samples

Derppening committed:
Remove debug message

Derppening committed:
WatBackend: Handle TypeRef in fieldSelect

Derppening committed:
WatBackend: Keep using TypeRef ID during signature generation

Derppening committed:
WatBackend: Scalafmt Reformat

Derppening committed:
WatBackend: Include this when getting local types

Derppening committed:
WasmGenerator: Add `global.{get,set}`

Derppening committed:
WatBackend: Implement resolution of local variables

Derppening committed:
Merge remote-tracking branch 'upstream/hkmc2' into wasm-backend

Derppening committed:
Apply scalafmt

Derppening committed:
WatBackend: Simplify generation of class definition

Derppening committed:
WatBackend: Fix redundant generation of returningTerm

Derppening committed:
WatBackend: Add HandleBlock handling

Derppening committed:
WIP - Implement struct.{set,get}

LPTK committed:
Make misc. improvements for `return`

# Conflicts:
#	hkmc2/shared/src/main/scala/hkmc2/semantics/Elaborator.scala
#	hkmc2/shared/src/test/mlscript/HkScratch.mls
#	hkmc2/shared/src/test/mlscript/HkScratch8.mls

LPTK committed:
Improve error reporting for non-local `return`

LPTK committed:
Fix location of type applications

Derppening committed:
WatBackend: Complete implementation of field assignment

Derppening committed:
wasm/Basics.mls: Add class defn, remove method call todo

Derppening committed:
Add `ref.null` instruction

Derppening committed:
WatBackend: Allocate globals during preamble

Derppening committed:
WatBackend: WIP - Implementing global Assign()

Derppening committed:
WatBackend: Fix globals WAT format

Derppening committed:
WasmGenerator: Fix documentation

Derppening committed:
WatBackend: Implement Select() resolution for static field access

Derppening committed:
WatBackend: Implement static field selection

Derppening committed:
Update Runtime.mjs

Derppening committed:
WatBackend: Add toRefExpr

Derppening committed:
wasm/Basics.mls: Add let-binding example

Derppening committed:
Merge remote-tracking branch 'upstream/hkmc2' into wasm-backend

Derppening committed:
WasmGenerator: Implement `ref.test`

Derppening committed:
WatBackend: Refactor and implement `+` operand matching for anyref

Derppening committed:
WIP: Initial commit for WatBuilder

Derppening committed:
WatBuilder: scalafmt

Derppening committed:
WatBuilder: Initial impl for boolean literals

Derppening committed:
WatBuilder: Minor WAT formatting fixes

Derppening committed:
WatBuilder: Add support for IntLit

Derppening committed:
Add CompilationTarget to Lowering.program()

Derppening committed:
WatBuilder: Refactor output for not-implemented IR

Derppening committed:
Merge remote-tracking branch 'upstream/hkmc2' into wasm-backend

Derppening committed:
WatBackend: Add WIP for class companion

Derppening committed:
WIP

Derppening committed:
WatBackend: Deletaget to WasmInstr where able

Derppening committed:
WatBuilder: Implement i32 addition

Derppening committed:
WIP

Derppening committed:
Fixup instruction types, delegate WatBackend instructions where possible

Derppening committed:
WatBuilder: Complete lowering of Assign()

Derppening committed:
WIP

Derppening committed:
WIP: Switch up some stuff

- Use Symbol as type/func/local keys - Collect and propagate locals to
function

Derppening committed:
WatBuilder: Always emit `elem` section for functions

Derppening committed:
WatBackend: Move all internal classes into ModuleProxy

Avoid name conflicts.

Derppening committed:
WatBuilder: Finish porting function defs and calls

Derppening committed:
WatBuilder: Add support for anonymous types

These types will be referenced by their numeric index instead.

Derppening committed:
WatBuilder: Update reasoning for funcref refinement

Derppening committed:
WatBuilder: Refactor to return tuple of WAT and main function name

Derppening committed:
Remove stale todo

Derppening committed:
WatBuilder: Remove `main` field

Derppening committed:
WatBuilder: Simplify extraction/matching of instruction prefix

Derppening committed:
WatBuilder: Require expressions to have at least one instruction

We have `nop` for that purpose.

Derppening committed:
Apply 100-col line limit

Derppening committed:
WatBuilder: Allocate entrypoint function name using Scope

Derppening committed:
Rename FuncRef to FuncId

For consistency with `Type{Ref,Id}`.

Derppening committed:
WatBuilder: Reimplement type/func handling

- Add FuncIdx to align with Type{Id,Idx} - Track all types in a single
Array, and split named types into Map - Add instance functions for
adding/getting types/functions - Fix erroneous call to returningTerm

Derppening committed:
Minor formatting changes

- Replace all usages of `dlof` with `fold` - Minor refactor using more
optional braces

Derppening committed:
Refactor indices

Derppening committed:
WatBuilder: Add var names generated by Scope to WAT

Derppening committed:
Merge remote-tracking branch 'upstream/hkmc2' into wasm-backend

Derppening committed:
Add missing changes from e4afd868875861ae471471af16e594848519d5de

Derppening committed:
Separate SignatureType from FunctionType

FunctionType is a fully-defined type, whereas SignatureType is just the
signature.

Derppening committed:
Update HeapType typedef

Derppening committed:
Generate ref.func in-place of function call site

Derppening committed:
Create (type) entry for functions and get use TypeIdx in ref.func expr
type

Derppening committed:
Add missing $ for param ID

Derppening committed:
Fix stack argument order for `call_ref`

Derppening committed:
Implement LocalIdx, parameter handling, lookup locals by symbolic name

Derppening committed:
Insert explicit TypeUse clause to function

Derppening committed:
Make `block` accept Seq[Result]

Better aligns with WAT.

Derppening committed:
Add guards for unsupported features for ClsLikeDefn

Derppening committed:
Merge remote-tracking branch 'upstream/hkmc2' into wasm-backend

Derppening committed:
Update warn/error locations to fix compilation errors

Derppening committed:
WatBuilder: WIP implementation for classes

Derppening committed:
Add non-Option function variants for `get{Type,Func}{,Info}`

Derppening committed:
WatBuilder: Refactor locals to be MutMap

Derppening committed:
Refactor FoldedInstr resultTypes

LPTK committed:
Changes from meeting

Derppening committed:
Update WatBuilder.scala

Derppening committed:
Update Basics.mls

Derppening committed:
WIP

Derppening committed:
Update WatBuilder.scala

Derppening committed:
Implement constructor generation

Derppening committed:
Pre-downcast Value.This before returning from result

The type is already known by that point.

Derppening committed:
Switch to matching field by field ident

Derppening committed:
Implement struct.get

Derppening committed:
Update binaryen

Derppening committed:
[DO NOT MERGE] Fix ValDefn

Derppening committed:
Use more specific symbol for lastWords(...)

Derppening committed:
Add workaround for incorrectly resolved symbols

Derppening committed:
Remove WatBackend

Weeeeeeee

Derppening committed:
WIP - Refactoring after removal of WatBackend

Derppening committed:
WIP - More refactoring after removal of WatBackend

Derppening committed:
Cleanup remaining unused WatBackend stuff, minor style cleanup

Derppening committed:
More reformatting

Derppening committed:
Reflow lines to 100col limit, more simplification

Derppening committed:
Merge remote-tracking branch 'upstream/hkmc2' into wasm-backend

Derppening committed:
Fix compilation errors post-merge

Derppening committed:
Pin Binaryen to latest stable release

We don't need nightly anymore.

LPTK commented:
Cool!

A couple preliminary things, here:

First, you shouldn't output the wasm code by default. It should only be output for _one or two_ representative blocks, and that's it. Otherwise, this will create way too much diff in git upon chnaging things.

Second,

> * `WasmGenerator.scala` and `BinaryenJSBackend.scala` are not needed in the actual implementation, and can be deleted if not useful.

If they're not used, let's not include them, as they'd be a net negative by making maintenance and evolution of the compiler more tedious. You can put them in a git tag and save the tag in some wiki/readme somewhere.

Derppening committed:
Remove WAT output

Derppening committed:
Remove BinaryenJSBackend, WasmGenerator, refactor WasmType to Type

Derppening commented:
> Cool!
> 
> A couple preliminary things, here:
> 
> First, you shouldn't output the wasm code by default. It should only be output for _one or two_ representative blocks, and that's it. Otherwise, this will create way too much diff in git upon chnaging things.

I am not sure which examples are "representative", so I just removed all of them for now ([fb02504](https://github.com/hkust-taco/mlscript/pull/341/commits/fb025047267400983115ad236539119e0a09cfc9)). Please let me know which examples you'd like the Wasm code and I will add the `:wat` directive back in.

> Second,
> 
> > * `WasmGenerator.scala` and `BinaryenJSBackend.scala` are not needed in the actual implementation, and can be deleted if not useful.
> 
> If they're not used, let's not include them, as they'd be a net negative by making maintenance and evolution of the compiler more tedious. You can put them in a git tag and save the tag in some wiki/readme somewhere.

I have saved the tag as `archive/binaryenjs-backend` and removed the unused files ([2d03019](https://github.com/hkust-taco/mlscript/pull/341/commits/2d030190ba9415e32d080b3598e53370eccfad72)).

Edit: I have also added the Wiki page.

LPTK commented on hkmc2DiffTests/src/test/scala/hkmc2/WasmDiffMaker.scala:
```diff
@@ -68,6 +68,9 @@ abstract class WasmDiffMaker extends LlirDiffMaker:
       given Raise =
+        // Ignore error about importing symbols not implemented
+        case err @ ErrorReport(source = Source.Compilation)
+            if "Import of symbol `.*Predef.mjs` not implemented yet".r.matches(err.mainMsg) => ()
```
WHAT

Derppening replied:
Addressed in 28d302a.

LPTK commented on hkmc2DiffTests/src/test/scala/hkmc2/WasmDiffMaker.scala:
```diff
@@ -103,12 +106,10 @@ abstract class WasmDiffMaker extends LlirDiffMaker:
           .replace('\n', ' ') |> host.execute match
           case ReplHost.Result(content) =>
-            output(content)
+            output(s"Wasm => $content")
```
```suggestion
            output(s"Wasm result: $content")
```

Derppening replied:
Addressed in 8579095 and 3cb5b78. Results are now outputted as (for example):

```
//│ Wasm result:
//│ = 1
```

LPTK commented on hkmc2DiffTests/src/test/scala/hkmc2/WasmDiffMaker.scala:
```diff
@@ -103,12 +106,10 @@ abstract class WasmDiffMaker extends LlirDiffMaker:
           .replace('\n', ' ') |> host.execute match
           case ReplHost.Result(content) =>
-            output(content)
+            output(s"Wasm => $content")
```
Can we pass this through our rendering function, just like what we do for JS results? What does it print, then? Can it enumerate the fields?

Derppening replied:
I am not sure if or how it would work. We can pass the result of `await wasm.binaryenRunFunc(...)` to the rendering function, but I am not quite sure if that is what you're asking for, nor do I know if it would be able to print anything meaningful.

LPTK replied:
To be clear, we're currently seeing NodeJS's own pretty-printing of the return value, correct? Then why not just pretty-print it ourselves, like we do with the other return values? You could actually achieve this by just making `binaryenRunFunc` call `Runtime`'s `render` function.

Derppening replied:
I just remembered this comment exists :sweat_smile: Addressed in 34ea2d0 by adding a parameter to `Wasm.binaryenRunFunc` which indicates whether the resulting object should be passed into `Runtime.render`.

Edit: It seems like the render function is unable to render the object correctly - it returns `{}` for a Wasm-created object :thinking:

LPTK replied:
Why do we get these strange quotes in the output? Maybe `render` is actually the wrong thing to call. (We want to call the same thing as we do for other results.) IIRC it might be something like `renderTopLevel`.

Derppening replied:
Fixed in eaef022 by adding `wasm.binaryenPrintFuncRes` that calls `Predef.print` under-the-hood which eliminates all the quotation marks.

Derppening committed:
Remove stupid workaround for Runtime.mjs import failure

Derppening committed:
WasmDiffMaker: Stop processing WAT when error is reported during
compilation

Derppening committed:
Run wasm compiled code by default when `:wasm` is set

Derppening committed:
Update Wasm execution result heading

Derppening committed:
Update Wasm result output format

Derppening committed:
Do not propagate UnreachableType

Derppening committed:
Add initial support for globals

Derppening committed:
Fix result type of constructor block

Derppening committed:
Initial support for passing importObject from outside

Derppening committed:
Rename Module.scala to Wasm.scala

Derppening committed:
Remove wasm.test function

Derppening committed:
Remove Num handling for wasm.plus_impl

Derppening committed:
Minor formatting changes

Derppening committed:
Update end markers

LPTK committed:
Merge branch 'hkmc2' into wasm-backend

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/codegen/Lowering.scala:
```diff
@@ -368,6 +369,9 @@ class Lowering()(using Config, TL, Raise, State, Ctx):
       case bs: BlockMemberSymbol =>
         bs.defn match
+        case S(_) if bs.asCls.exists(_ is ctx.builtins.Int31) =>
+        // case S(_) if bs.asCls.exists(sym => (sym is ctx.builtins.Int31) || (sym is ctx.builtins.wasm.Int31)) =>
```
Comment is obsolete?

Derppening replied:
Removed in 49aa8a2.

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/MLsCompiler.scala:
```diff
@@ -93,7 +94,7 @@ class MLsCompiler(preludeFile: os.Path, mkOutput: ((Str => Unit) => Unit) => Uni
       val jsb = ltl.givenIn:
         codegen.js.JSBuilder()
-      val le = low.program(blk)
+      val le = low.program(blk, CompilationTarget.JS)
```
I think this belongs in the `Config` object, which you can pass here implicitly.

Derppening replied:
Addressed in f1855ae, although I am not too sure if it is correct to set the default target as JS.

LPTK replied:
Probably fine for now.

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/codegen/Lowering.scala:
```diff
@@ -434,8 +438,21 @@ class Lowering()(using Config, TL, Raise, State, Ctx):
         conclude(Value.Ref(State.runtimeSymbol).selN(Tree.Ident("try_catch")))
+      case t if t.resolvedSym.exists(_ is ctx.builtins.wasm.plus_impl) =>
+        conclude(Value.Ref(State.runtimeSymbol).selN(Tree.Ident("plus_impl")))
+      // case t if t.resolvedSym.exists(sym => (sym is ctx.builtins.Int31) || (sym is ctx.builtins.wasm.Int31)) =>
```
Comment is obsolete?

Derppening replied:
Removed in 49aa8a2.

Derppening committed:
Remove commented `wasm.Int31` builtins

Derppening committed:
Move CompilationTarget to Config.scala

Derppening committed:
Update tests after removing `target` from `Program`

Derppening commented:
Hmm... I am not quite sure why the CI suddenly failed like this. I tried cloning my branch from scratch, running `hkmc2AllTests/test` using `sbt` and that works...

LPTK commented on hkmc2DiffTests/src/test/scala/hkmc2/MLsDiffMaker.scala:
```diff
@@ -54,6 +54,11 @@ abstract class MLsDiffMaker extends DiffMaker:
+  /**
+   * Enables Wasm support. All options in [[WasmDiffMaker]] are no-op if this option is not set.
+   */
+  val wasm = NullaryCommand("wasm")
```
Any reason for moving this definition from `WasmDiffMaker.scala`? It seemed fine there.

Derppening replied:
It's to determine whether to set `Config.target` as `Wasm` or `JS` (see `MLsDiffMaker.scala:96`).

On second thought, I think I can override `mkConfig` in `WasmDiffMaker` and create a new `Config` instance there (i.e. `super.mkConfig.copy(target = ...)`, what do you think about that?

LPTK replied:
Oh no, overriding would make the logic 200% harder to understand for no upside at all. The current version is fine.

LPTK replied:
An alternative would be to move the definition back to `WasmDiffMaker` but use:
```scala
abstract class MLsDiffMaker extends DiffMaker:
  this: WasmDiffMaker =>
```

Derppening replied:
I tried applying your suggestion and got this:

```
[error] -- [E049] Reference Error: /home/david/programming/mlscript/hkmc2DiffTests/src/test/scala/hkmc2/MLsDiffMaker.scala:256:4
[error] 256 |    processTerm(e, inImport = false)
[error]     |    ^^^^^^^^^^^
[error]     |    Reference to processTerm is ambiguous.
[error]     |    It is both defined in class MLsDiffMaker
[error]     |    and inherited subsequently in class MLsDiffMaker
[error]     |
[error]     | longer explanation available when compiling with `-explain`
```

Changing `processTerm` to `this.processTerm` causes this error in `JSBackendDiffMaker.scala`:

```
[error] -- [E058] Type Mismatch Error: /home/david/programming/mlscript/hkmc2DiffTests/src/test/scala/hkmc2/JSBackendDiffMaker.scala:19:15
[error] 19 |abstract class JSBackendDiffMaker extends MLsDiffMaker:
[error]    |               ^
[error]    |illegal inheritance: self type hkmc2.JSBackendDiffMaker of class JSBackendDiffMaker does not conform to self type hkmc2.WasmDiffMaker
[error]    |of parent class MLsDiffMaker
[error]    |
[error]    | longer explanation available when compiling with `-explain`
```

I think I will keep the current version for now, unless you have more suggestions on how to fix this issue.

LPTK commented on hkmc2DiffTests/src/test/scala/hkmc2/WasmDiffMaker.scala:
```diff
@@ -0,0 +1,108 @@
+      s"const $wasmSuppNme = (await import(\"${wasmSuppFile}\")).default;"
+    ) match
+      case ReplHost.Result(msg) =>
+        if msg.startsWith("Uncaught") then
```
```suggestion
        if msg.startsWith(ReplHost.uncaughtErrorHead) then
```

Derppening replied:
I got this error when making this change:

```
[error] -- [E173] Reference Error: /home/david/programming/mlscript/hkmc2DiffTests/src/test/scala/hkmc2/WasmDiffMaker.scala:36:35
[error] 36 |        if msg.startsWith(ReplHost.uncaughtErrorHead) then
[error]    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]    |value uncaughtErrorHead cannot be accessed as a member of hkmc2.ReplHost.type from class WasmDiffMaker.
[error]    |  private value uncaughtErrorHead can only be accessed from object ReplHost in package hkmc2.
```

Seems like `ReplHost.uncaughtErrorHead` is a `private val` in `ReplHost`.

LPTK replied:
I think you can un-private it.

Derppening replied:
Addressed in 7df4271.

LPTK commented:
> Hmm... I am not quite sure why the CI suddenly failed like this. I tried cloning my branch from scratch, running `hkmc2AllTests/test` using `sbt` and that works...

Weird, it works on my computer, too. Maybe the CI is using an outdated version of Node. Can you try to request a newer one in the YAML file?

Derppening commented:
> > Hmm... I am not quite sure why the CI suddenly failed like this. I tried cloning my branch from scratch, running `hkmc2AllTests/test` using `sbt` and that works...
> 
> Weird, it works on my computer, too. Maybe the CI is using an outdated version of Node. Can you try to request a newer one in the YAML file?

Do you mean changing `nodejs_22` in `flake.nix` to a newer version? I don't see a way to bump the version of Node from `.github/workflows/nix.yml`.

LPTK commented:
> Do you mean changing `nodejs_22` in `flake.nix` to a newer version? I don't see a way to bump the version of Node from `.github/workflows/nix.yml`.

IDK, there must be a way, I was just randomly suggesting a likely location.

Derppening committed:
Use `ReplHost.uncaughtHostHead` for message prefix matching

Derppening committed:
Bump Node.js to v24 in Nix

Derppening committed:
Update Nix dependencies

Nixpkgs is updated to 25.05.

Derppening commented:
> > Do you mean changing `nodejs_22` in `flake.nix` to a newer version? I don't see a way to bump the version of Node from `.github/workflows/nix.yml`.
> 
> IDK, there must be a way, I was just randomly suggesting a likely location.

I tried bumping NixOS to 25.05 and Node to v24, but it's still failing for some reason...

Derppening committed:
WIP - Test with `:wat` and `:showRepl`

Derppening committed:
Merge branch 'hkmc2' into wasm-backend

Derppening committed:
Fixup after Document refactoring and remove debugging flags

Derppening commented:
So it turns out it's the changes from 9a2651f that changed how `Document.toString` works, which caused the amazing diff of

```

 //│ Wat:
-//│ (module
-//│   (type (func (result (ref null any))))
-//│   (func $entry (type 0) (result (ref null any))
-//│     (ref.i31
-//│       (i32.const 0)))
-//│   (export "entry" (func $entry))
-//│   (elem declare func $entry))
+//│ DocCons(DocCons(DocText((module),DocNest(2,DocCons(DocCons(DocCons(DocCons(DocText(),DocBreak(false)),DocText()),DocCons(DocCons(DocCons(DocCons(DocCons(DocText((type ),DocText()),DocText()),DocCons(DocCons(DocText((func),DocCons(DocText( ),DocCons(DocCons(DocText((result ),DocCons(DocCons(DocCons(DocCons(DocText((ref),DocText( null)),DocText( )),DocText(any)),DocText()))),DocText())))),DocText()))),DocText())),DocCons(DocCons(DocCons(DocText(),DocBreak(false)),DocText()),DocCons(DocCons(DocCons(DocCons(DocCons(DocCons(DocCons(DocCons(DocCons(DocCons(DocText((func ),DocCons(DocCons(DocText($),DocText(entry)),DocText())),DocText( (type )),DocCons(DocCons(DocText(),DocText(0)),DocText())),DocText())),DocCons(DocText( ),DocCons(DocCons(DocText((result ),DocCons(DocCons(DocCons(DocCons(DocText((ref),DocText( null)),DocText( )),DocText(any)),DocText()))),DocText())))),DocText()),DocNest(2,DocCons(DocCons(DocCons(DocCons(DocCons(DocText(),DocText()),DocBreak(false)),DocText()),DocCons(DocCons(DocCons(DocCons(D
```

And because the CI tests the version of my PR merged into master (and our local branches haven't been updated against master), it seemed to work on our machines :sweat_smile: 

It's all fixed now, but I guess there's no harm in keeping the Nixpkgs/Node version bump changes? Or do you prefer for me to split this to a separate PR?

LPTK commented:
> the CI tests the version of my PR merged into master

Ah yeah, it's not the first time this bites us... Can be very surprising behavior. There should at least be a warning when the CI runs on a commit that doesn't exist in the PR 🙄 

Also, I'm amazed by how uninformative and useless the error messages of a high-profile industrial tool can be.

> It's all fixed now, but I guess there's no harm in keeping the Nixpkgs/Node version bump changes?

👍

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/codegen/Lowering.scala:
```diff
@@ -1004,7 +1019,6 @@ class Lowering()(using Config, TL, Raise, State, Ctx):
         val l = new TempSymbol(N)
         Assign(l, r, k(l |> Value.Ref.apply))
   
-  
```
?

Derppening replied:
Restored in 4c75c22.

LPTK commented on hkmc2/shared/src/test/mlscript-compile/Runtime.mls:
```diff
@@ -1,3 +1,6 @@
+// TODO: Implement this in Lowering of MLsCompiler
+// wasm
```
Not sure what these comments are saying and why they're here.

Derppening replied:
The comment is removed in e0efb64. Please refer to [this reply](https://github.com/hkust-taco/mlscript/pull/341#discussion_r2409929091) for why they are here.

LPTK commented on hkmc2/shared/src/test/mlscript/wasm/Basics.mls:
```diff
@@ -0,0 +1,66 @@
+//│ = [Object: null prototype] {}
+
+
+class Foo(val a)
```
Let's show the wasm for this one.

Derppening replied:
Done in 4a264cf.

Derppening replied:
I mistakenly added the flag to the wrong function, actually done in 044dade.

LPTK commented on hkmc2/shared/src/test/mlscript/wasm/Basics.mls:
```diff
@@ -0,0 +1,66 @@
+//│ = 42
+
+
+fun foo() = 42
```
Let's show the wasm for this one.

Derppening replied:
Done in 4a264cf.

Derppening replied:
Same as above - Actually done in 044dade.

LPTK commented on hkmc2DiffTests/src/test/scala/hkmc2/MLsDiffMaker.scala:
```diff
@@ -87,7 +92,8 @@ abstract class MLsDiffMaker extends DiffMaker:
       )),
-      liftDefns = Opt.when(liftDefns.isSet)(LiftDefns())
+      liftDefns = Opt.when(liftDefns.isSet)(LiftDefns()),
+      if wasm.isSet then CompilationTarget.Wasm else CompilationTarget.JS,
```
Is this actually used? I can't see where.

Derppening replied:
This is supposed to be used when `wasm` is declared in the top of an `mls` file (see [this comment](https://github.com/hkust-taco/mlscript/pull/341#discussion_r2409541282)). I can remove this if you want this change to be incorporated once the relevant changes are made.

LPTK commented on hkmc2DiffTests/src/test/scala/hkmc2/MLsDiffMaker.scala:
```diff
@@ -87,7 +92,8 @@ abstract class MLsDiffMaker extends DiffMaker:
       )),
-      liftDefns = Opt.when(liftDefns.isSet)(LiftDefns())
+      liftDefns = Opt.when(liftDefns.isSet)(LiftDefns()),
+      if wasm.isSet then CompilationTarget.Wasm else CompilationTarget.JS,
```
Let's name this argument, just like the other ones.

Derppening replied:
Done in 9b20256.

LPTK reviewed (approved):
Ok, nice work! Let's merge this when the remaining comments are addressed (I'll do the merge).

Derppening committed:
Restore missing newline

Derppening committed:
Remove `wasm` comment in Runtime.mls

Derppening committed:
Fix more Document.{toString => mkString} migrations

Derppening committed:
Show WAT for function and class example

Derppening committed:
Add argument name to Config constructor

Derppening committed:
Remove more redundant `wasm.Int31` in Preludes

Derppening committed:
Add parameter to `Wasm.binaryenRunFunc` for passing result to `render()`

Derppening committed:
Fix brainfart

LPTK committed:
Merge branch 'hkmc2' into wasm-backend

Derppening committed:
Remove redundant Promise for binaryenFmtWat

Derppening committed:
Add `binaryenPrintFuncRes` and use `Predef.print` for wasm outputs

LPTK commented on hkmc2/shared/src/test/mlscript-compile/Wasm.mls:
```diff
@@ -37,11 +39,14 @@ fun binaryenCompileToModule(wat, importObject): Promise =
       func(wasmMod.instance.exports)
     )
+
+fun binaryenPrintFuncRes(wat, importObject, func, render): Promise = 
```
What's the point of having a `render` parameter?

Derppening replied:
It's to determine whether to pass the resulting object to `Runtime.render` before printing the output. I can remove it if there is no reason why we want to print an object without pretty-printing first.

Derppening replied:
Removed the `render` parameter in be54b28. We can add this back further down the line if it is needed.

LPTK replied:
Yea, there was literally no point.

Derppening committed:
Remove render parameter from `binaryenPrintFuncRes`

LPTK committed:
Merge branch 'hkmc2' into wasm-backend
